update open_dataset to display dataset names

bradyrx opened the pull request:
# Description

This removes the function `get_dataset()` and instead returns the available datasets when `open_dataset()` is passed with no arguments.

Fixes https://github.com/bradyrx/climpred/issues/121

## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?
![Screen Shot 2019-05-13 at 2 37 21 PM](https://user-images.githubusercontent.com/8881170/57652375-a70c8e80-758c-11e9-8ace-992f0a8b0755.png)

ahuang11 commented on climpred/loadutils.py:
```diff
@@ -105,48 +73,57 @@ def open_dataset(name, cache=True, cache_dir=_default_cache_dir,
+            print(f"'{key}': {file_descriptions[key]}")
+
+    if name is None:
+        get_datasets()
```
A trick to lessen indentation downstream is to `return` immediately in the function.

bradyrx replied:
Good point. The indentation felt weird.

ahuang11 commented on climpred/loadutils.py:
```diff
@@ -105,48 +73,57 @@ def open_dataset(name, cache=True, cache_dir=_default_cache_dir,
+    def get_datasets():
+        """Prints out available datasets for the user to load if no args are
+           given."""
+        for key in file_descriptions.keys():
```
I'm wondering if we should use upper case for `file_descriptions` and `file_alias_dict` since both are a constant?
https://www.python.org/dev/peps/pep-0008/#constants

```
FILE_ALIAS_DICT = {'MPI-control-1D': 'PM_MPI-ESM-LR_control',
                   'MPI-control-3D': 'PM_MPI-ESM-LR_control3d',
                   'MPI-PM-DP-1D': 'PM_MPI-ESM-LR_ds',
                   'MPI-PM-DP-3D': 'PM_MPI-ESM-LR_ds3d',
                   'CESM-DP-SST': 'CESM-DP-LE.SST.global',
                   'CESM-DP-SSS': 'CESM-DP-LE.SSS.global',
                   'CESM-LE': 'CESM-LE.global_mean.SST.1955-2015',
                   'ERSST': 'ERSSTv4.global.mean',
                   'FOSI-SST': 'FOSI.SST.global',
                   'FOSI-SSS': 'FOSI.SSS.global'}

file_descriptions = {'MPI-PM-DP-1D': 'decadal prediction ensemble area' +
                                     ' averages of SST/SSS/AMO.',
                     'MPI-PM-DP-3D': 'decadal prediction ensemble' +
                                     ' lat/lon/time of SST/SSS/AMO.',
                     'MPI-control-1D': 'area averages for the control run of' +
                                       ' SST/SSS.',
                     'MPI-control-3D': 'lat/lon/time for the control run of' +
                                       ' SST/SSS.',
                     'CESM-DP-SST': 'decadal prediction ensemble of global' +
                                    ' mean SSTs.',
                     'CESM-DP-SSS': 'decadal prediction ensemble of global' +
                                    ' mean SSS.',
                     'CESM-LE': 'uninitialized ensemble of global mean SSTs.',
                     'ERSST': 'observations of global mean SSTs.',
                     'FOSI-SST': 'reconstruction of global mean SSTs.',
                     'FOSI-SSS': 'reconstruction of global mean SSS.',
                     }
```
Also, I have mixed feelings about nested functions, but it doesn't really matter.

ahuang11 commented on climpred/loadutils.py:
```diff
@@ -105,48 +73,57 @@ def open_dataset(name, cache=True, cache_dir=_default_cache_dir,
+        get_datasets()
+    else:
+        if name.endswith('.nc'):
+            name = name[:-3]
```
Maybe use
name, ext = os.path.splitext(name), which you can reuse ext on line 91 and if we ever have other file types besides `.nc` like `.grb` or something.
https://stackoverflow.com/questions/541390/extracting-extension-from-filename-in-python

bradyrx replied:
This is helpful. I think other split methods I was using would split at every period, which I had within the file name. But this separates the extension from the filename.

ahuang11 commented on climpred/loadutils.py:
```diff
@@ -105,48 +73,57 @@ def open_dataset(name, cache=True, cache_dir=_default_cache_dir,
+                msg = """
+                MD5 checksum does not match, try downloading dataset again.
+                """
+                raise IOError(msg)
```
I wonder if we could implement an automatic remove, download, retry method here for the people like me who are clueless about where this is downloaded? Or simply a more specific msg about the path and instructions?

bradyrx replied:
This code is hacked together from the `xarray` tutorial code so this might be just residual from there. I see what you mean though. I'll update with a more specific message for now.

bradyrx replied:
I don't fully understand the md5 files either honestly.

bradyrx commented:
Okay all issues resolved. Force-pushed some squashed commits. Will merge after Travis runs. Thanks for the nitpicky stuff - I think it's important to learn these best practices.

bradyrx commented:
Oops I broke the whole function. Need to dig into this.

bradyrx commented:
I'm confused here. Can @aaronspring or @ahuang11 try checking out this pull request and trying to download datasets? It's failing now and not due to code changes... just won't download any data properly.

See this Travis fail just now on master: https://travis-ci.com/bradyrx/climpred/builds/111789632

I'm not sure if Travis runs on my server somehow (I figured it runs on its own server). I'm concerned that this is a proxy issue because I'm on a LANL computer where remote retrieval is weird sometimes.

aaronspring commented:
doesnt work for me.

```python
>>> open_dataset('FOSI-SSS',cache=False)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/aaron.spring/Coding/climpred/climpred/loadutils.py", line 122, in open_dataset
    _urlretrieve(url, localfile)
  File "/Users/aaron.spring/anaconda3/envs/xr/lib/python3.7/urllib/request.py", line 247, in urlretrieve
    with contextlib.closing(urlopen(url, data)) as fp:
  File "/Users/aaron.spring/anaconda3/envs/xr/lib/python3.7/urllib/request.py", line 222, in urlopen
    return opener.open(url, data, timeout)
  File "/Users/aaron.spring/anaconda3/envs/xr/lib/python3.7/urllib/request.py", line 531, in open
    response = meth(req, response)
  File "/Users/aaron.spring/anaconda3/envs/xr/lib/python3.7/urllib/request.py", line 641, in http_response
    'http', request, response, code, msg, hdrs)
  File "/Users/aaron.spring/anaconda3/envs/xr/lib/python3.7/urllib/request.py", line 569, in error
    return self._call_chain(*args)
  File "/Users/aaron.spring/anaconda3/envs/xr/lib/python3.7/urllib/request.py", line 503, in _call_chain
    result = func(*args)
  File "/Users/aaron.spring/anaconda3/envs/xr/lib/python3.7/urllib/request.py", line 649, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 404: Not Found

bradyrx commented:
I'm really at a loss here. I migrated all of the data over to http://github.com/bradyrx/climpred-data. I re-ran the checksum script.

Running the following I get a dataset, so it isn't a network issue at LANL.
```python
from xarray.tutorial import open_dataset
open_dataset('air_temperature')
```

Then running:
```python
open_dataset('FOSI.SSS.global', github_url='https://github.com/bradyrx/climpred-data')
```

I get the following two errors:
```shell
gaierror: [Errno 8] nodename nor servname provided, or not known
URLError: <urlopen error [Errno 8] nodename nor servname provided, or not known>
```

This breaks at the line `urlretrieve(url, localfile)`. Although I have set the url variable to output, and I get `https://github.com/bradyrx/climpred-data/raw/master/FOSI.SSS.global.nc`, which when put into my browser downloads the file.

No idea why our respositories specifically do not work. They are public. Any ideas??

ahuang11 commented:
Hehe, try printing out the url and you'll notice it doesn't have an ext because os.path.splitext detected .global as the ext...

`https://github.com/bradyrx/climpred/raw/master/sample_data/prediction/FOSI.SSS.global`

My suggestion: add `.nc` to all the filenames here; that way if we add a grib example dataset, we'll know the difference.
```
FILE_ALIAS_DICT = {'MPI-control-1D': 'PM_MPI-ESM-LR_control',
                   'MPI-control-3D': 'PM_MPI-ESM-LR_control3d',
                   'MPI-PM-DP-1D': 'PM_MPI-ESM-LR_ds',
                   'MPI-PM-DP-3D': 'PM_MPI-ESM-LR_ds3d',
                   'CESM-DP-SST': 'CESM-DP-LE.SST.global',
                   'CESM-DP-SSS': 'CESM-DP-LE.SSS.global',
                   'CESM-LE': 'CESM-LE.global_mean.SST.1955-2015',
                   'MPIESM_miklip_baseline1-hind-SST-global':
                   'MPIESM_miklip_baseline1-hind-SST-global',
                   'MPIESM_miklip_baseline1-hist-SST-global':
                   'MPIESM_miklip_baseline1-hist-SST-global',
                   'MPIESM_miklip_baseline1-assim-SST-global':
                   'MPIESM_miklip_baseline1-assim-SST-global',
                   'ERSST': 'ERSSTv4.global.mean',
                   'FOSI-SST': 'FOSI.SST.global',
                   'FOSI-SSS': 'FOSI.SSS.global'}
```

Or a quick fix right now would be reverting back to:
```
    if not ext.endswith('.nc'):
        ext += '.nc'
```

bradyrx commented:
Man, I swear I checked that a bunch, plus the xarray version wasn't working which doesn't use `splitext()`. Anyways, I fixed that and updated the pointer to our new `climpred-data` repo and it worked. Tested all keys in `FILE_ALIAS_DICT`.

After it goes through Travis, I'll remove our data directory here and squash commits and merge. We can open another PR to rename as you said (and as @aaronspring brought up in https://github.com/bradyrx/climpred/issues/116).

Can also try to address https://github.com/bradyrx/climpred/issues/126 in the next PR.

bradyrx commented:
Okay it passed the checks now. Gonna squash commits and remove the data from this repo now that it lives on `climpred-data`.

bradyrx commented:
Okay, testing pulls from `climpred-data` now and checks that our aliases are up-to-date/it can load in all expected sample data. Will check Travis results in morning and hopefully merge...

bradyrx commented:
Okay this one is in. Need to follow up with https://github.com/bradyrx/climpred/issues/126 and maybe move to underscore_naming for the climpred-data repo.
